### DISCUSSION TITLE
Build and run a local per-project `sea-orm-cli`?

### DISCUSSION THREAD

**Discussion by Expurple** [SeaQL/sea-orm#1889] [Q&A] [answered]
Currently, [the recommended way](https://www.sea-ql.org/sea-orm-tutorial/ch01-02-migration-cli.html#initialize-using-sea-orm-cli) to install `sea-orm-cli` is to install it globally using `cargo intall sea-orm-cli`. This approach ignores the `sea-orm` version pinned in my project's `Cargo.lock` and blindly installs the latest version. This affects the generated code and results in large diffs. I had to manually `cargo intall sea-orm-cli@MY.VER.SION`. But this is not a good workaround, it doesn't track changes in my `Cargo.toml` and doesn't scale to multiple projects with different `sea-orm` versions.

**Is there a way to build and run `sea-orm-cli` locally, `npx`-style?**

This would be the best solution for me. I want to track `sea-orm-cli` in the project's `Cargo.lock` and keep it local. But I haven't found any way to do so. Have you explored the relevant features of `cargo`? If you've researched this too and `cargo` doesn't support this, I can create a feature request for `cargo`.

**Comment by tyt2y3** [comment, marked as answer]
That's a legit setup. I would suggest you to always upgrade to the latest version, but meanwhile before you upgrade, you can:

1) create a crate in the workspace that depends on a particular version of `sea-orm-cli`
2) create a main.rs and call the cli commands
3) advice your teammates to use the local version, i.e. `my-project-cli`

Ref: https://github.com/SeaQL/sea-orm/blob/master/sea-orm-cli/src/bin/main.rs

There's a tiny bit of boilerplate, but everything is exposed as `lib` in the cli.

**Reply by Expurple** [reply]
> I would suggest you to always upgrade to the latest version

Sure, but sometimes I'm just working on a PR that isn't supposed to change that. And by the way, if I use the global latest `sea-orm-cli` during that, can't it in theory generate code that's incompatible with the project's `sea-orm` version? That seems like a real downside to your current approach. Another downside is that the global version doesn't auto-update when the project's `sea-orm` version is updated. I really think that using a local synchronized version is the right way.

To get rid of the boilerplate in the future, I'll make a feature request for `cargo` as I planned to. Meanwhile, can you also export the whole `main()` function? That would save me an unnecessary repeated copy-paste in case `sea-orm-cli` is updated and something's changed there.

**Reply by Expurple** [reply]
My [cargo feature request](https://github.com/rust-lang/cargo/issues/12764).

**Reply by Expurple** [reply]
I prepared a pull request with exported `main()` function: #2034, it would make applying your workaround easier
